webextension/background.js

```diff
@@ -129,8 +127,76 @@ const Config = (function() {
       }
     }
 
-    load() {
+    maybeSendTelemetry(message) {
+      browser.study.sendTelemetry(message);
+    }
+
+    get shieldStudySetup() {
+      const pref = `extensions.${browser.runtime.id.split("@")[0]}.variation`;
+      return {
+        activeExperimentName: browser.runtime.id,
+        allowEnroll: true,
+        studyType: "shield",
+        telemetry: {
+          send: true,
+          removeTestingFlag: !this._testingMode,
+        },
+        endings: {
+          "user-disable": {
+            baseUrl: null,
+          },
+          expired: {
+            baseUrl: null,
+          },
+          dataPermissionsRevoked: {
+            baseUrl: null,
+            study_state: "ended-neutral",
+          },
+        },
+        logLevel: this._testingMode ? 30 : 0,
+        variationOverridePreference: pref,
+        weightedVariations: UIVariants.map(name => {
+          return {name, weight: 1};
+        }),
+        expire: {
+          days: 60,
+        },
+      };
+    }
+
+    _activateShield() {
+      if (this._shieldActivatedPromise) {
+        return this._shieldActivatedPromise;
+      }
+
+      this._shieldActivatedPromise = new Promise((resolve, reject) => {
+        const endListener = studyInfo => {
+          browser.study.onReady.removeListener(readyListener);
+          browser.study.onEndStudy.removeListener(endListener);
+          this._shieldActivatedPromise = undefined;
+          reject(studyInfo);
+        };
+        const readyListener = studyInfo => {
+          browser.study.onReady.removeListener(readyListener);
+          browser.study.onEndStudy.removeListener(endListener);
+          this._shieldActivatedPromise = undefined;
+          resolve(studyInfo);
+        };
+        browser.study.onReady.addListener(readyListener);
+        browser.study.onEndStudy.addListener(endListener);
+        try {
+          browser.study.setup(this.shieldStudySetup).catch(reject);
+        } catch (err) {
+          endListener();
+          reject(err);
```

Comment:
This line also looks interesting...
If we somehow end up in this `catch{}` block, we manually call `endListener()` which removes some listeners, and rejects the Promise chain with a `studyInfo` object, which would be `undefined`. But then _this_ line will also reject the promise, this time with an error object. So it looks like our Promise may potentially reject twice.

Reply:
Yeah, I think there's still a slight variation, as one path will reject with an event-like object, and the other path will reject with an error-like object. But I guess it all depends on how the rejection is being used/handled.